web/modules/custom_weather_module/src/Plugin/Block/CustomWeatherBlock.php

```diff
@@ -89,9 +100,52 @@ private function userWeather() {
     }
     $location = $this->userLocation();
     $client = new Client();
-    $request = $client->get('http://api.weatherapi.com/v1/current.json?key=' . $api_key . '&q=' . $location);
-    $response = json_decode($request->getBody(), TRUE);
-    return $response['current']['temp_c'] . ' °C, ' . $response['current']['condition']['text'];
+    try {
+      $request = $client->get('http://api.weatherapi.com/v1/current.json?key=' . $api_key . '&q=' . $location);
+      $response = json_decode($request->getBody(), TRUE);
+      return [
+        'conditions' => $response['current']['temp_c'] . ' °C, ' . $response['current']['condition']['text'],
+        'icon' => $response['current']['condition']['icon'],
+      ];
+    }
+    catch (RequestException $e) {
+      // @todo use dependency injection
+      // phpcs:ignore
+      \Drupal::logger('custom_weather_module')->error('Message from @module: @message', [
+        '@module' => 'custom_weather_module',
+        '@message' => "GuzzleHttp\Exception\ConnectException: cURL error 6: Could not resolve host: weatherapi.com",
+      ]);
+    }
+
+  }
+
+  /**
+   * Private function that gets data from cache or sets it.
+   */
+  private function userWeather() {
+    $cacheId = 'weatherCache';
+    // @todo use dependency injection
+    // phpcs:ignore
+    if ((\Drupal::cache()->get($cacheId))
+      // phpcs:ignore
+      && (\Drupal::cache()->get($cacheId)->data['location'] === $this->userLocation())
+      // phpcs:ignore
+      && (\Drupal::cache()->get($cacheId)->data['conditions'])) {
+      // @todo use dependency injection
+      // phpcs:ignore
+      return \Drupal::cache()->get($cacheId)->data;
+    }
+    else {
+      $data = [
+        'location' => $this->userLocation(),
+        'conditions' => $this->getWeather()['conditions'],
+        'icon' => $this->getWeather()['icon'],
+      ];
+      // @todo use dependency injection
+      // phpcs:ignore
+      \Drupal::cache()->set($cacheId, $data,time() + 3600);
+      return $this->getWeather();
```

Comment:
There are three calling of $this->getWeather(). The better practice is to save the result of $this->getWeather() to a variable and then use that variable three times instead.